server/src/main/java/org/elasticsearch/transport/SniffConnectionStrategy.java

```diff
@@ -299,31 +301,53 @@ private void collectRemoteNodes(Iterator<Supplier<DiscoveryNode>> seedNodesSuppl
-                    // existing context information.
-                    threadContext.markAsSystemContext();
-                    transportService.sendRequest(
-                        connection,
-                        ClusterStateAction.NAME,
-                        request,
-                        TransportRequestOptions.EMPTY,
-                        responseHandler
-                    );
+
+                if (REMOTE_CLUSTER_PROFILE.equals(connectionManager.getConnectionProfile().getTransportProfile())) {
+                    TransportService.ContextRestoreResponseHandler<RemoteClusterNodesAction.Response> responseHandler =
+                        new TransportService.ContextRestoreResponseHandler<>(
+                            threadContext.newRestorableContext(false),
+                            new RemoteClusterNodesSniffResponseHandler(connection, listener, seedNodesSuppliers)
+                        );
+                    try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+                        // we stash any context here since this is an internal execution and should not leak any
+                        // existing context information.
+                        threadContext.markAsSystemContext();
+                        transportService.sendRequest(
+                            connection,
+                            RemoteClusterNodesAction.NAME,
+                            RemoteClusterNodesAction.Request.INSTANCE,
+                            TransportRequestOptions.EMPTY,
+                            responseHandler
+                        );
+                    }
+                } else {
+                    ClusterStateRequest request = new ClusterStateRequest();
+                    request.clear();
+                    request.nodes(true);
+                    TransportService.ContextRestoreResponseHandler<ClusterStateResponse> responseHandler =
+                        new TransportService.ContextRestoreResponseHandler<>(
+                            threadContext.newRestorableContext(false),
+                            new ClusterStateSniffResponseHandler(connection, listener, seedNodesSuppliers)
+                        );
+                    try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+                        // we stash any context here since this is an internal execution and should not leak any
+                        // existing context information.
+                        threadContext.markAsSystemContext();
+                        transportService.sendRequest(
+                            connection,
+                            ClusterStateAction.NAME,
+                            request,
+                            TransportRequestOptions.EMPTY,
+                            responseHandler
+                        );
+                    }
```

Comment:
Optional but the following should work and cut down on a bit of code redundancy: 

```java
final String action;
final TransportRequest request;
final AbstractSniffResponseHandler<?> sniffResponseHandler;
if (REMOTE_CLUSTER_PROFILE.equals(connectionManager.getConnectionProfile().getTransportProfile())) {
    action = RemoteClusterNodesAction.NAME;
    request = RemoteClusterNodesAction.Request.INSTANCE;
    sniffResponseHandler = new RemoteClusterNodesSniffResponseHandler(connection, listener, seedNodesSuppliers);
} else {
    action = ClusterStateAction.NAME;
    ClusterStateRequest innerRequest = new ClusterStateRequest();
    innerRequest.clear();
    innerRequest.nodes(true);
    request = innerRequest;
    sniffResponseHandler = new ClusterStateSniffResponseHandler(connection, listener, seedNodesSuppliers);
}
TransportService.ContextRestoreResponseHandler<?> responseHandler = new TransportService.ContextRestoreResponseHandler<>(
    threadContext.newRestorableContext(false),
    sniffResponseHandler
);
try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
    // we stash any context here since this is an internal execution and should not leak any
    // existing context information.
    threadContext.markAsSystemContext();
    transportService.sendRequest(connection, action, request, TransportRequestOptions.EMPTY, responseHandler);
}
```

Reply:
I was trying to not touch the existing remote cluster logic. But that's probably overly cautious. I adopted your suggestion. Thanks!